packages/penrose-core/src/compiler/Style.ts

```diff
@@ -350,8 +354,10 @@ const checkRelPattern = (varEnv: Env, rel: RelationPattern): StyErrors => {
 
     // TODO(error)
     if (isErr(res1)) {
-      // TODO(error): extract and return error -- how to convert from SubstanceError to StyError?
-      return ["substance typecheck error in B"];
+      const subErr1: SubstanceError = unsafelyGetErr(res1);
+      // TODO(error): Do we need to wrap this error further, or is returning SubstanceError with no additional Style info ok?
+      // return ["substance typecheck error in B"];
+      return [{ tag: "TaggedSubstanceError", error: subErr1 }];
```

Comment:
Ideally, I think you'll need to add the source locations of the relevant selector clauses as well. I can make `checkVar` return a more specific type and you can add a wrapper around those types and customize the error messages further if needed.

Reply:
> Ideally, I think you'll need to add the source locations of the relevant selector clauses as well. 

Yeah, the right thing to do is to pass in the selector environments, which have the locations.

> I can make checkVar return a more specific type and you can add a wrapper around those types and customize the error messages further if needed.

Not sure what you mean by this (but thanks for the offer); I feel like the above should suffice.

Reply:
Okay. Remember to fix the `unsafelyGetErr`. We can revisit this if it ends up harder to read for the users.